File: ios/Sources/MoQKit/MoQPlayer.swift
import AVFoundation
import CoreMedia
import MoQKitFFI

// MARK: - PlaybackStats

/// A snapshot of playback quality metrics, sampled over the most recent one-second window.
///
/// Obtain the current snapshot via ``MoQPlayer/stats``, which is safe to call from any
/// thread or actor without holding the main actor.
public struct PlaybackStats: Sendable {
    /// Estimated end-to-end audio latency in milliseconds (wall-clock delay from sender to speaker).
    /// `nil` when no audio track is active.
    public let audioLatencyMs: Double?
    /// Estimated end-to-end video latency in milliseconds.
    /// `nil` when no video track is active.
    public let videoLatencyMs: Double?

    /// Audio stall statistics since playback started. `nil` when no audio track is active.
    public let audioStalls: StallStats?
    /// Video stall statistics since playback started. `nil` when no video track is active.
    public let videoStalls: StallStats?

    /// Audio bitrate of the incoming stream in kilobits per second. `nil` when no audio track is active.
    public let audioBitrateKbps: Double?
    /// Video bitrate of the incoming stream in kilobits per second. `nil` when no video track is active.
    public let videoBitrateKbps: Double?

    /// Time from ``MoQPlayer/play()`` to the first decoded audio frame, in milliseconds.
    /// `nil` before the first frame arrives or when no audio track is active.
    public let timeToFirstAudioFrameMs: Double?
    /// Time from ``MoQPlayer/play()`` to the first decoded video frame, in milliseconds.
    /// `nil` before the first frame arrives or when no video track is active.
    public let timeToFirstVideoFrameMs: Double?

    /// Current video frame rate in frames per second. `nil` when no video track is active.
    public let videoFps: Double?

    /// Total audio frames dropped since playback started. `nil` when no audio track is active.
    public let audioFramesDropped: UInt64?
    /// Total video frames dropped since playback started. `nil` when no video track is active.
    public let videoFramesDropped: UInt64?

    /// Current audio ring buffer fill level in milliseconds. Reflects how much audio is
    /// buffered ahead of the playhead. `nil` when no audio track is active.
    public let audioRingBufferMs: Double?
    /// Current video jitter buffer fill level in milliseconds. `nil` when no video track is active.
    public let videoJitterBufferMs: Double?
}

/// Stall statistics for a single track since playback started.
public struct StallStats: Sendable {
    /// Number of stall events (playback interruptions) since playback started.
    public let count: UInt64
    /// Total cumulative duration of all stall events in milliseconds.
    public let totalDurationMs: Double
    /// Fraction of playback time spent stalling: `totalDurationMs / totalPlaybackDurationMs`.
    public let rebufferingRatio: Double
}

// MARK: - MoQPlayerEvent

/// Events emitted on ``MoQPlayer/events`` during playback.
public enum MoQPlayerEvent: Sendable {
    /// The first frame of the given track kind was successfully decoded and rendered.
    case trackPlaying(TrackKind)
    /// Playback of the given track kind was paused via ``MoQPlayer/pause()``.
    case trackPaused(TrackKind)
    /// The remote sender stopped the given track (the track stream ended).
    case trackStopped(TrackKind)
    /// All active tracks have stopped. The events stream completes immediately after this event.
    case allTracksStopped
    /// A non-fatal error occurred on the given track. Playback of other tracks continues.
    case error(TrackKind, String)

    /// Identifies which type of media track an event relates to.
    public enum TrackKind: String, Sendable {
        case video, audio
    }
}

// MARK: - MoQPlayer

/// Adaptive real-time player for MoQ media streams.
///
/// `MoQPlayer` subscribes to one or two tracks from a ``MoQBroadcastInfo`` (one video and/or
/// one audio track), decodes the incoming frames, and renders them in sync:
///
/// - Video frames are rendered into ``videoLayer`` — an `AVSampleBufferDisplayLayer` you can
///   embed in any `UIView` or `CALayer` hierarchy.
/// - Audio frames are decoded to PCM and played through `AVAudioEngine` using the system's
///   default audio output. No additional audio session configuration is required.
///
/// ```swift
/// let player = try MoQPlayer(tracks: broadcastInfo.videoTracks + broadcastInfo.audioTracks)
/// view.layer.addSublayer(player.videoLayer)
/// try await player.play()
/// ```
///
/// The class is `@MainActor` — all calls must be made from the main actor.
@MainActor
public final class MoQPlayer {
    /// The `AVSampleBufferDisplayLayer` that receives decoded video frames.
    ///
    /// Add this layer to your view hierarchy before calling ``play()``.
    public let videoLayer: AVSampleBufferDisplayLayer
    /// Emits ``MoQPlayerEvent`` values as playback progresses.
    ///
    /// The stream completes after ``allTracksStopped`` is emitted or after ``stopAll()`` is called.
    public let events: AsyncStream<MoQPlayerEvent>

    private let tracks: [any MoQTrackInfo]
    private var targetBufferingMs: UInt64
    private let eventsContinuation: AsyncStream<MoQPlayerEvent>.Continuation

    private var audioRenderer: AudioRenderer?
    private var videoRenderer: VideoRenderer?

    private var videoSubscription: MoQMediaTrack?
    private var audioSubscription: MoQMediaTrack?

    private var videoTask: Task<Void, Never>?
    private var audioTask: Task<Void, Never>?
    private var coordinatorTask: Task<Void, Never>?

    nonisolated(unsafe) private var audioTracer: PacketTimingTracer?
    nonisolated(unsafe) private var videoTracer: PacketTimingTracer?
    nonisolated(unsafe) private var audioRendererForStats: AudioRenderer?
    nonisolated(unsafe) private var videoRendererForStats: VideoRenderer?

    private let accumulator = PlaybackMetricsAccumulator()
    private let mode: Mode

    private enum Mode {
        case audioVideo
        case audioOnly
        case videoOnly
    }

    private nonisolated var hasVideoTrack: Bool {
        tracks.contains(where: { $0 is MoQVideoTrackInfo })
    }
    private nonisolated var hasAudioTrack: Bool {
        tracks.contains(where: { $0 is MoQAudioTrackInfo })
    }

    /// Creates a player for the given tracks.
    ///
    /// - Parameters:
    ///   - tracks: One or two ``MoQTrackInfo`` values from ``MoQBroadcastInfo``. Pass at most one
    ///     video track and one audio track. Mixing two video or two audio tracks is not supported.
    ///   - targetBufferingMs: Target playout delay in milliseconds. Higher values improve
    ///     resilience to network jitter at the cost of increased end-to-end latency. Defaults
    ///     to 100 ms. Can be adjusted live via ``updateTargetLatency(ms:)``.
    /// - Throws: ``MoQSessionError/invalidConfiguration(_:)`` if `tracks` is empty or contains
    ///   more than two entries.
    public init(
        tracks: [any MoQTrackInfo],
        targetBufferingMs: UInt64 = 100
    ) throws {
        if tracks.isEmpty || tracks.count > 2 {
            throw MoQSessionError.invalidConfiguration("expected one or two tracks")
        }

        self.tracks = tracks
        self.targetBufferingMs = targetBufferingMs
        self.videoLayer = AVSampleBufferDisplayLayer()

        let hasVideo = tracks.contains(where: { $0 is MoQVideoTrackInfo })
        let hasAudio = tracks.contains(where: { $0 is MoQAudioTrackInfo })
        if hasVideo && hasAudio {
            mode = .audioVideo
        } else if hasAudio {
            mode = .audioOnly
        } else {
            mode = .videoOnly
        }

        var cont: AsyncStream<MoQPlayerEvent>.Continuation!
        self.events = AsyncStream { cont = $0 }
        self.eventsContinuation = cont
    }

    // MARK: - Public API

    /// Adjusts the target playout delay without interrupting playback.
    ///
    /// The change takes effect immediately on both the audio ring buffer and the video jitter
    /// buffer. Lowering the value reduces latency but increases the risk of stalls on lossy
    /// networks; raising it improves resilience.
    ///
    /// - Parameter ms: New target buffering depth in milliseconds.
    public func updateTargetLatency(ms: UInt64) {
        targetBufferingMs = ms
        audioRenderer?.updateTargetLatency(ms: Int(ms))
        videoRenderer?.updateTargetBuffering(ms: ms)
    }

    /// A snapshot of current playback quality metrics.
    ///
    /// Safe to call from any thread or actor — does not require the main actor.
    /// Values are sampled over the most recent one-second window. See ``PlaybackStats``
    /// for field-level documentation.
    public nonisolated var stats: PlaybackStats {
        accumulator.snapshot(
            audioLatencyMs: hasAudioTrack ? audioTracer?.latencyMs : nil,
            videoLatencyMs: hasVideoTrack ? videoTracer?.latencyMs : nil,
            audioRingBufferMs: audioRendererForStats?.bufferFillMs,
            videoJitterBufferMs: videoRendererForStats?.bufferFillMs
        )
    }

    /// Subscribes to the selected tracks and begins decoding and rendering.
    ///
    /// Playback events are emitted on ``events``. Call ``pause()`` to temporarily suspend
    /// rendering without releasing the track subscriptions, or ``stopAll()`` to fully tear down.
    ///
    /// - Throws: ``MoQSessionError`` if a track subscription or renderer initialisation fails.
    public func play() async throws {
        guard videoTask == nil && audioTask == nil else { return }

        try subscribe()

        let timebase = try Self.createTimebase()

        if hasAudioTrack {
            audioTracer = PacketTimingTracer(kind: .audio, timebase: timebase) { report in
                MoQLogger.player.debug("\(report)")
            }
        }
        if hasVideoTrack {
            videoTracer = PacketTimingTracer(kind: .video, timebase: timebase) { report in
                MoQLogger.player.debug("\(report)")
            }
        }

        accumulator.markPlayStart()

        try setupAudioRenderer(timebase: timebase)
        try setupVideoRenderer(timebase: timebase)

        startIngestTasks()
    }

    /// Pauses playback and cancels all track subscriptions.
    ///
    /// Emits ``MoQPlayerEvent/trackPaused(_:)`` for each active track. To resume, call
    /// ``play()`` again — it will re-subscribe to the tracks and restart rendering from the
    /// current live position.
    public func pause() async {
        teardown(permanent: false)

        if hasVideoTrack {
            eventsContinuation.yield(.trackPaused(.video))
        }
        if hasAudioTrack {
            eventsContinuation.yield(.trackPaused(.audio))
        }
    }

    /// Stops playback, closes all track subscriptions, and releases all rendering resources.
    ///
    /// The ``events`` stream completes after this call. The player cannot be reused — create a
    /// new ``MoQPlayer`` instance to start playback again.
    public func stopAll() async {
        MoQLogger.player.debug("Stopping real-time player")
        teardown(permanent: true)
    }

    deinit {
        videoTask?.cancel()
        audioTask?.cancel()
        coordinatorTask?.cancel()
        eventsContinuation.finish()
    }

    // MARK: - Private: teardown

    private func teardown(permanent: Bool) {
        videoTask?.cancel()
        audioTask?.cancel()
        coordinatorTask?.cancel()
        videoTask = nil
        audioTask = nil
        coordinatorTask = nil

        audioRenderer?.stop()
        videoRenderer?.stop()
        videoRenderer?.flush()

        audioTracer?.reset()
        videoTracer?.reset()

        videoSubscription?.close()
        audioSubscription?.close()
        videoSubscription = nil
        audioSubscription = nil

        if permanent {
            audioRenderer = nil
            videoRenderer = nil
            audioTracer = nil
            videoTracer = nil
            audioRendererForStats = nil
            videoRendererForStats = nil
            accumulator.reset()

            eventsContinuation.finish()
        }
    }

    // MARK: - Private: play() helpers

    private nonisolated static func createTimebase() throws -> CMTimebase {
        var tb: CMTimebase?
        CMTimebaseCreateWithSourceClock(
            allocator: kCFAllocatorDefault,
            sourceClock: CMClockGetHostTimeClock(),
            timebaseOut: &tb
        )
        guard let tb else {
            throw MoQSessionError.invalidConfiguration("Failed to create CMTimebase")
        }
        CMTimebaseSetTime(tb, time: .zero)
        CMTimebaseSetRate(tb, rate: 0)
        return tb
    }

    private func setupAudioRenderer(timebase: CMTimebase) throws {
        guard let aInfo = tracks.compactMap({ $0 as? MoQAudioTrackInfo }).first else {
            return
        }

        let renderer = try AudioRenderer(
            config: aInfo.rawConfig,
            timebase: timebase,
            targetLatencyMs: Int(targetBufferingMs),
            metrics: accumulator
        )
        try renderer.start()
        self.audioRenderer = renderer
        self.audioRendererForStats = renderer
    }

    private func setupVideoRenderer(timebase: CMTimebase) throws {
        guard let vInfo = tracks.compactMap({ $0 as? MoQVideoTrackInfo }).first else {
            return
        }

        let renderer = try VideoRenderer(
            config: vInfo.rawConfig,
            timebase: timebase,
            isTimebaseOwner: mode == .videoOnly,
            targetBufferingMs: targetBufferingMs,
            layer: videoLayer,
            metrics: accumulator
        )
        renderer.start()
        self.videoRenderer = renderer
        self.videoRendererForStats = renderer
    }

    private func startIngestTasks() {
        let continuation = eventsContinuation
        let audioTracer = self.audioTracer
        let videoTracer = self.videoTracer
        let metrics = self.accumulator

        // Audio ingest task
        if let aTrack = audioSubscription, let renderer = audioRenderer {
            audioTask = Task.detached {
                var lastPtsUs: UInt64 = 0
                var firstFrame = true
                for await frame in aTrack.frames {
                    if Task.isCancelled { break }
                    do {
                        if Self.isDiscontinuity(
                            currentUs: frame.timestampUs, lastUs: lastPtsUs,
                            keyframe: frame.keyframe
                        ) {
                            MoQLogger.player.debug(
                                "Audio discontinuity detected, flushing")
                            renderer.flush()
                            audioTracer?.reset()
                        }
                        lastPtsUs = frame.timestampUs

                        metrics.recordAudioBytes(frame.payload.count)
                        audioTracer?.record(ptsUs: frame.timestampUs)
                        let pcm = try renderer.decoder.decode(payload: frame.payload)
                        renderer.enqueue(pcm: pcm, timestampUs: frame.timestampUs)

                        if firstFrame {
                            firstFrame = false
                            metrics.markFirstAudioFrame()
                            continuation.yield(.trackPlaying(.audio))
                        }
                    } catch {
                        MoQLogger.player.error("Audio decode error: \(error)")
                        continuation.yield(.error(.audio, error.localizedDescription))
                    }
                }
                if !Task.isCancelled {
                    continuation.yield(.trackStopped(.audio))
                }
            }
        }

        // Video ingest task
        if let vTrack = videoSubscription, let renderer = videoRenderer, renderer.canProcess {
            videoTask = Task.detached {
                var lastPtsUs: UInt64 = 0
                var firstFrame = true
                for await frame in vTrack.frames {
                    if Task.isCancelled { break }
                    do {
                        if Self.isDiscontinuity(
                            currentUs: frame.timestampUs, lastUs: lastPtsUs,
                            keyframe: frame.keyframe
                        ) {
                            MoQLogger.player.debug(
                                "Video discontinuity detected, flushing")
                            renderer.flush()
                            videoTracer?.reset()
                        }
                        lastPtsUs = frame.timestampUs

                        metrics.recordVideoBytes(frame.payload.count)
                        videoTracer?.record(ptsUs: frame.timestampUs)
                        let inserted = try renderer.insert(
                            payload: frame.payload, timestampUs: frame.timestampUs,
                            keyframe: frame.keyframe)

                        if inserted && firstFrame {
                            firstFrame = false
                            metrics.markFirstVideoFrame()
                            continuation.yield(.trackPlaying(.video))
                        }
                    } catch {
                        MoQLogger.player.error("Video frame processing error: \(error)")
                        continuation.yield(.error(.video, error.localizedDescription))
                    }
                }
                if !Task.isCancelled {
                    continuation.yield(.trackStopped(.video))
                }
            }
        }

        // Coordinator: wait for both tasks and emit allTracksStopped
        let vTask = videoTask
        let aTask = audioTask
        coordinatorTask = Task.detached {
            await vTask?.value
            await aTask?.value
            if !Task.isCancelled {
                continuation.yield(.allTracksStopped)
                continuation.finish()
            }
        }
    }

    // MARK: - Private: helpers

    /// Detects a PTS discontinuity: keyframe with >500ms jump from the last timestamp.
    private nonisolated static func isDiscontinuity(
        currentUs: UInt64, lastUs: UInt64, keyframe: Bool
    ) -> Bool {
        guard keyframe && lastUs > 0 else { return false }
        let diff =
            currentUs > lastUs
            ? currentUs - lastUs
            : lastUs - currentUs
        return diff > 500_000
    }

    private func subscribe() throws {
        for track in tracks {
            if let vInfo = track as? MoQVideoTrackInfo {
                MoQLogger.player.debug(
                    "Video track: \(vInfo.name), codec=\(vInfo.config.codec), config=\(vInfo.config.debugDescription), container=\(vInfo.rawConfig.container)"
                )
                do {
                    videoSubscription = try MoQMediaTrack(
                        broadcast: vInfo.broadcast, name: vInfo.name,
                        container: vInfo.rawConfig.container,
                        maxLatencyMs: targetBufferingMs)
                } catch {
                    MoQLogger.player.error(
                        "Failed to subscribe to video track \(vInfo.name): \(error)")
                }
            } else if let aInfo = track as? MoQAudioTrackInfo {
                MoQLogger.player.debug(
                    "Audio track: \(aInfo.name), config = \(aInfo.config.debugDescription), container=\(aInfo.rawConfig.container)"
                )
                do {
                    audioSubscription = try MoQMediaTrack(
                        broadcast: aInfo.broadcast, name: aInfo.name,
                        container: aInfo.rawConfig.container,
                        maxLatencyMs: targetBufferingMs)
                } catch {
                    MoQLogger.player.error(
                        "Failed to subscribe to audio track \(aInfo.name): \(error)")
                }
            }
        }
    }

}

File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/MoQPlayer.kt
package com.swmansion.moqkit

import android.util.Log
import android.view.Surface
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.launch

private const val TAG = "MoQRealTimePlayer"

/**
 * Real-time audio+video player with fine-grained latency control.
 *
 * Uses MediaCodec (async) for decoding, AudioTrack (MODE_STREAM) for audio output,
 * and a Surface-configured MediaCodec for video output — bypassing ExoPlayer for lower latency.
 *
 * ### Typical usage
 * ```kotlin
 * val player = MoQPlayer(broadcastInfo.videoTracks + broadcastInfo.audioTracks,
 *                        targetLatencyMs = 150, parentScope = lifecycleScope)
 * player.setSurface(surfaceView.holder.surface)
 * player.play()
 * // … later …
 * player.stop()
 * ```
 *
 * @param tracks List of tracks to play. Both [MoQAudioTrackInfo] and [MoQVideoTrackInfo] are
 *   optional — any combination works, including video-only or audio-only.
 * @param targetLatencyMs Target end-to-end playback latency in milliseconds. Lower values
 *   reduce delay at the cost of increased risk of stalls. Defaults to 100 ms.
 * @param parentScope Coroutine scope whose lifetime bounds the player's internal coroutines.
 */
class MoQPlayer(
    private val tracks: List<MoQTrackInfo>,
    private val targetLatencyMs: Int = 100,
    parentScope: CoroutineScope,
) {
    init {
        val audioTracksCount = tracks.count { track ->  track is MoQAudioTrackInfo }
        if (audioTracksCount > 1) {
            throw IllegalArgumentException("at most one audio track is allowed")
        }
        val videoTracksCount = tracks.count { track -> track is MoQVideoTrackInfo }
        if (videoTracksCount > 1) {
            throw IllegalArgumentException("at most one video track is allowed")
        }
        if (audioTracksCount + videoTracksCount == 0) {
            throw IllegalArgumentException("at least one audio or video track is expected")
        }
    }

    /**
     * Playback lifecycle events emitted by the player.
     */
    sealed class Event {
        /** A track started producing decoded output. [kind] is `"audio"` or `"video"`. */
        data class TrackPlaying(val kind: String) : Event()
        /** A track was paused via [pause]. [kind] is `"audio"` or `"video"`. */
        data class TrackPaused(val kind: String) : Event()
        /** A track's incoming data stream ended. [kind] is `"audio"` or `"video"`. */
        data class TrackStopped(val kind: String) : Event()
        /** An unrecoverable error occurred on a track. */
        data class Error(val kind: String, val message: String) : Event()
        /** All active tracks have stopped (stream ended or [stop] called). */
        object AllTracksStopped : Event()
    }

    private val scope = CoroutineScope(parentScope.coroutineContext + SupervisorJob())

    private val _events = MutableSharedFlow<Event>(extraBufferCapacity = 8)
    val events: SharedFlow<Event> = _events

    private var audioRenderer: AudioRenderer? = null
    private var videoRenderer: VideoRenderer? = null
    private var audioIngestJob: Job? = null
    private var videoIngestJob: Job? = null
    private var surface: Surface? = null
    private var playing = false
    private val accumulator = PlaybackMetricsAccumulator()

    /** Current playback time in microseconds. */
    val currentTimeUs: Long get() = audioRenderer?.currentTimeUs ?: 0L

    /** Snapshot of current playback metrics. */
    val stats: PlaybackStats get() {
        val timeUs = audioRenderer?.currentTimeUs ?: 0L
        val audioLatency = audioRenderer?.let {
            if (it.lastIngestPtsUs > 0 && timeUs > 0) (it.lastIngestPtsUs - timeUs).toDouble() / 1000.0 else null
        }
        val videoLatency = videoRenderer?.let {
            if (it.lastIngestPtsUs > 0 && timeUs > 0) (it.lastIngestPtsUs - timeUs).toDouble() / 1000.0 else null
        }
        return accumulator.snapshot(
            audioLatencyMs = audioLatency,
            videoLatencyMs = videoLatency,
            audioRingBufferMs = audioRenderer?.bufferFillMs,
            videoJitterBufferMs = videoRenderer?.bufferFillMs,
        )
    }

    /**
     * Set or clear the video output surface.
     * If play() was already called and a surface becomes available, starts the video pipeline.
     * If surface becomes null, stops the video renderer.
     */
    fun setSurface(surface: Surface?) {
        this.surface = surface
        if (surface != null && playing && videoRenderer == null) {
            startVideo(surface)
        } else if (surface == null && videoRenderer != null) {
            stopVideo()
        }
    }

    /**
     * Starts audio (and video, if a surface is set) playback.
     *
     * Opens subscriptions for all tracks in the list provided at construction time and begins
     * decoding. The first decoded audio frame triggers an [Event.TrackPlaying] event.
     * Safe to call if a surface has not yet been provided — video will start automatically
     * once [setSurface] is called. Audio is optional — if no audio track is present only
     * video is played.
     */
    fun play() {
        playing = true
        accumulator.markPlayStart()

        startAudio()

        surface?.let {
            startVideo(it)
        }
    }

    private fun startAudio() {
        val audioInfo = tracks.filterIsInstance<MoQAudioTrackInfo>().firstOrNull() ?: run {
            Log.d(TAG, "No audio track, skipping audio pipeline")
            return
        }

        Log.d(TAG, "startAudio: '${audioInfo.name}' ${audioInfo.config.sampleRate}Hz " +
            "${audioInfo.config.channelCount}ch, targetLatency=${targetLatencyMs}ms")

        val renderer = AudioRenderer(
            config = audioInfo.config,
            targetLatencyMs = targetLatencyMs,
            metrics = accumulator,
        )
        audioRenderer = renderer
        renderer.start()

        val audioFlow = subscribeTrack(
            audioInfo.broadcast,
            audioInfo.name,
            audioInfo.config.container,
            targetLatencyMs.toULong(),
        )

        audioIngestJob = scope.launch {
            var firstFrame = true
            try {
                audioFlow.collect { frame ->
                    // TODO: add discontinuation detection
                    renderer.submitFrame(frame.payload, frame.timestampUs.toLong())
                    accumulator.recordAudioBytes(frame.payload.size)

                    if (firstFrame) {
                        firstFrame = false
                        accumulator.markFirstAudioFrame()
                        Log.d(TAG, "First audio frame received")
                        _events.tryEmit(Event.TrackPlaying("audio"))
                    }
                }

                _events.tryEmit(Event.TrackStopped("audio"))
            } catch (e: Exception) {
                Log.e(TAG, "Audio ingest error: $e")
                _events.tryEmit(Event.Error("audio", e.message ?: "Unknown error"))
            }
            checkAllStopped()
        }
    }

    private fun startVideo(surface: Surface) {
        val videoInfo = tracks.filterIsInstance<MoQVideoTrackInfo>().firstOrNull() ?: return

        Log.d(TAG, "Starting video: '${videoInfo.name}' codec=${videoInfo.config.codec}")

        val track = VideoRendererTrack(videoInfo.config, targetLatencyMs.toLong() * 1000)

        val renderer = VideoRenderer(
            activeTrack = track,
            surface = surface,
            timebase = audioRenderer?.timebase,
            metrics = accumulator,
        )
        videoRenderer = renderer
        renderer.start()

        videoIngestJob = launchVideoIngestJob(videoInfo, track)
    }

    private fun launchVideoIngestJob(videoInfo: MoQVideoTrackInfo, track: VideoRendererTrack): Job {
        Log.d(TAG, "Subscribing to video track")
        val videoFlow = subscribeTrack(
            videoInfo.broadcast,
            videoInfo.name,
            videoInfo.config.container,
            targetLatencyMs.toULong(),
        )

        return scope.launch {
            var firstFrame = true
            try {
                Log.d(TAG, "Waiting for video frames")
                videoFlow.collect { frame ->
                    // TODO: add discontinuation detection

                    track.insert(frame.payload, frame.timestampUs.toLong(), frame.keyframe)
                    accumulator.recordVideoBytes(frame.payload.size)

                    if (firstFrame) {
                        firstFrame = false
                        accumulator.markFirstVideoFrame()
                        Log.d(TAG, "First video frame received")
                        _events.tryEmit(Event.TrackPlaying("video"))
                    }
                }

                _events.tryEmit(Event.TrackStopped("video"))
            } catch (e: Exception) {
                Log.e(TAG, "Video ingest error: $e")
                _events.tryEmit(Event.Error("video", e.message ?: "Unknown error"))
            }
            checkAllStopped()
        }
    }

    /**
     * Switches to a different video rendition seamlessly.
     *
     * Creates a pending track that accumulates frames in the background. The swap state
     * machine in [VideoRenderer] decides between a seamless cut-in and a flush-and-swap,
     * then calls [onActivated] on the renderer's HandlerThread.
     *
     * No-ops if a switch is already in progress.
     *
     * @param videoInfo The new video rendition to switch to.
     * @param onActivated Called on the renderer's HandlerThread when the swap completes.
     */
    fun switchVideoTrack(videoInfo: MoQVideoTrackInfo, onActivated: (() -> Unit)? = null) {
        val renderer = videoRenderer ?: return
        if (renderer.hasPendingTrack) return

        Log.d(TAG, "Switching video track to '${videoInfo.name}' codec=${videoInfo.config.codec}")

        val newTrack = VideoRendererTrack(videoInfo.config, targetLatencyMs.toLong() * 1000)
        val oldJob = videoIngestJob

        renderer.setPendingTrack(newTrack) {
            // TODO: we may encounter situation where the track will never be switched so we will leak the job
            oldJob?.cancel()
            onActivated?.invoke()
        }

        videoIngestJob = launchVideoIngestJob(videoInfo, newTrack)
    }

    private fun stopVideo() {
        videoIngestJob?.cancel()
        videoIngestJob = null
        videoRenderer?.stop()
        videoRenderer = null
    }

    private fun checkAllStopped() {
        val audioDone = audioIngestJob?.isActive != true
        val videoDone = videoIngestJob?.isActive != true
        if (audioDone && videoDone) {
            _events.tryEmit(Event.AllTracksStopped)
        }
    }

    /**
     * Pauses playback by stopping all decoders and cancelling ingest coroutines.
     *
     * Unlike [stop], the player can be resumed by calling [play] again. Emits
     * [Event.TrackPaused] for each active track.
     */
    fun pause() {
        Log.d(TAG, "pause")
        playing = false
        audioIngestJob?.cancel()
        audioIngestJob = null
        videoIngestJob?.cancel()
        videoIngestJob = null
        audioRenderer?.stop()
        audioRenderer = null
        videoRenderer?.stop()
        videoRenderer = null

        val hasAudio = tracks.any { it is MoQAudioTrackInfo }
        val hasVideo = tracks.any { it is MoQVideoTrackInfo }
        if (hasAudio) _events.tryEmit(Event.TrackPaused("audio"))
        if (hasVideo) _events.tryEmit(Event.TrackPaused("video"))
    }

    /**
     * Stops playback and resets all accumulated metrics.
     *
     * Releases decoders and cancels ingest coroutines. Call [play] to start again from scratch.
     */
    fun stop() {
        Log.d(TAG, "stop")
        playing = false
        audioIngestJob?.cancel()
        audioIngestJob = null
        videoIngestJob?.cancel()
        videoIngestJob = null
        audioRenderer?.stop()
        audioRenderer = null
        videoRenderer?.stop()
        videoRenderer = null
        accumulator.reset()
    }

    /**
     * Adjusts the target playback latency while the player is running.
     *
     * Propagates the new value to the audio ring buffer and video jitter buffer immediately.
     * Lower values cause more aggressive frame dropping to catch up to live; higher values
     * trade latency for smoother playback.
     *
     * @param ms New target latency in milliseconds.
     */
    fun updateTargetLatency(ms: Int) {
        audioRenderer?.updateTargetLatency(ms)
        videoRenderer?.updateTargetBuffering(ms)
    }
}
